pinot-core/src/main/java/org/apache/pinot/core/data/table/OrderedIndexedTableResizer.java

```diff
@@ -0,0 +1,236 @@
+      aggregationColumnToIndex.put(aggregationColumn, i);
+      aggregationColumnToInfo.put(aggregationColumn, aggregationInfo);
+    }
+
+    _numOrderBy = orderBy.size();
+    _orderByInfos = new ArrayList<>(_numOrderBy);
+    Comparator[] comparators = new Comparator[_numOrderBy];
+
+    for (int i = 0; i < _numOrderBy; i++) {
+      SelectionSort selectionSort = orderBy.get(i);
+      String column = selectionSort.getColumn();
+
+      if (keyIndexMap.containsKey(column)) {
+        int index = keyIndexMap.get(column);
+        _orderByInfos.add(createOrderByInfoForKeyColumn(index));
+      } else if (aggregationColumnToIndex.containsKey(column)) {
+        int index = aggregationColumnToIndex.get(column);
+        _orderByInfos.add(createOrderByInfoForAggregationColumn(index, aggregationColumnToInfo.get(column)));
+      } else {
+        throw new UnsupportedOperationException("Could not find column " + column + " in data schema");
+      }
+
+      comparators[i] = ComparableComparator.getInstance();
+      if (!selectionSort.isIsAsc()) {
+        comparators[i] = comparators[i].reversed();
+      }
+    }
+
+    _intermediateRecordComparator = (o1, o2) -> {
+
+      for (int i = 0; i < _numOrderBy; i++) {
+        int result = comparators[i].compare(o1.getValues()[i], o2.getValues()[i]);
+        if (result != 0) {
+          return result;
+        }
+      }
+      return 0;
+    };
+  }
+
+  /**
+   * Constructs an IntermediateRecord from Record
+   * The IntermediateRecord::key is the same Record::key
+   * The IntermediateRecord::values contains only the order by columns, in the query's sort sequence
+   * For aggregation values in the order by, the final result is extracted if the intermediate result is non-comparable
+   */
+  @VisibleForTesting
+  IntermediateRecord getIntermediateRecord(Record record) {
+    Comparable[] intermediateRecordValues = new Comparable[_numOrderBy];
+
+    Object[] keyColumns = record.getKey().getColumns();
+    Object[] aggregationColumns = record.getValues();
+    for (int i = 0; i < _numOrderBy; i++) {
+      OrderByInfo orderByInfo = _orderByInfos.get(i);
+      Comparable comparable;
+      if (orderByInfo.getOrderByType().equals(OrderByType.KEY_COLUMN)) {
+        Object keyColumn = keyColumns[orderByInfo.getIndex()];
+        comparable = (Comparable) keyColumn; // FIXME: is this the right way to get Comparable? will it work for BYTES?
+      } else {
+        Object aggregationColumn = aggregationColumns[orderByInfo.getIndex()];
+        AggregationFunction aggregationFunction = orderByInfo.getAggregationFunction();
+        if (!aggregationFunction.isIntermediateResultComparable()) {
+          comparable = aggregationFunction.extractFinalResult(aggregationColumn);
+        } else {
+          comparable = (Comparable) aggregationColumn;
+        }
+      }
+      intermediateRecordValues[i] = comparable;
+    }
+    return new IntermediateRecord(record.getKey(), intermediateRecordValues);
```

Comment:
Do we need to create this additional object for all cases (for example, when we don't need to extract final results)?

Reply:
True, we do not need to create it if we don't have to extract final result. However, the new IntermediateRecord would only have as many elements as the size of order by clause, which should be really small. It will only be alive during resize. The Key within IntermediateRecord is referring to the Key within the original Record (so no new Object there).
I'll add a TODO here, to re-evaluate this during further optimizations, and possibly avoid creating IntermediateRecord when possible. Right now, I'd prefer to avoid handling this, as it'll bring in a lot more if else.